sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorSender.java

```diff
@@ -264,53 +267,90 @@ public Mono<Void> send(List<Message> messageBatch, DeliveryState deliveryState)
+                    if (sectionBytes.length > 0) {
+                        totalEncodedSize += sectionBytes.length;
+                        if (totalEncodedSize > maxMessageSize) {
+                            return batchBufferOverflowError(maxMessageSize);
+                        }
+                        buffer.append(sectionBytes);
+                    }
                 }
 
-                return send(bytes, byteArrayOffset, AmqpConstants.AMQP_BATCH_MESSAGE_FORMAT, deliveryState);
+                return send(buffer, AmqpConstants.AMQP_BATCH_MESSAGE_FORMAT, deliveryState);
             }).then();
     }
 
+    private byte[] batchEnvelopBytes(Message envelopMessage, int maxMessageSize) {
+        // Proton-j doesn't support multiple dataSections to be part of AmqpMessage.
+        // Here's the alternate approach provided: https://github.com/apache/qpid-proton/pull/54
+        final Message message = Proton.message();
+        message.setMessageAnnotations(envelopMessage.getMessageAnnotations());
+        final int size = messageSerializer.getSize(message);
+        final int allocationSize = Math.min(size + MAX_AMQP_HEADER_SIZE_BYTES, maxMessageSize);
+        final byte[] encodedBytes = new byte[allocationSize];
+        final int encodedSize = message.encode(encodedBytes, 0, allocationSize);
+        // This copyOf copying is just few bytes for envelop.
+        return Arrays.copyOf(encodedBytes, encodedSize);
+    }
+
+    private byte[] batchBinaryDataSectionBytes(Message sectionMessage, int maxMessageSize) {
+        final int size = messageSerializer.getSize(sectionMessage);
+        final int allocationSize = Math.min(size + MAX_AMQP_HEADER_SIZE_BYTES, maxMessageSize);
+        final byte[] encodedBytes = new byte[allocationSize];
+        final int encodedSize = sectionMessage.encode(encodedBytes, 0, allocationSize);
+
+        final Message message = Proton.message();
+        final Data binaryData = new Data(new Binary(encodedBytes, 0, encodedSize));
+        message.setBody(binaryData);
+        final int binaryRawSize = binaryData.getValue().getLength();
+        // Precompute the "amqp:data:binary" encoded size -
+        final int binaryEncodedSize = binaryEncodedSize(binaryRawSize);
+        // ^ this pre-computation avoids allocating byte[] 'arr1' of estimated encoded size (to pass to message.encode(arr1,))
+        // and a second allocation of byte[] 'arr2' with exact encoded size (returned from message.encode(arr1,)) then
+        // copying encoded size bytes from 'arr1' to 'arr2'. Skipping extra allocations and CPU cycles for copying.
+        final byte[] binaryEncodedBytes = new byte[binaryEncodedSize];
+        message.encode(binaryEncodedBytes, 0, binaryEncodedSize);
+        return binaryEncodedBytes;
+    }
+
+    private Mono<Void> batchBufferOverflowError(int maxMessageSize) {
+        return Mono.error(new AmqpException(false, AmqpErrorCondition.LINK_PAYLOAD_SIZE_EXCEEDED,
```

Comment:
nit: should we also add a log record with [monoError](https://github.com/Azure/azure-sdk-for-java/blob/527b28b969a1bea28970e7d3b54a10c2cdca8503/sdk/core/azure-core/src/main/java/com/azure/core/util/FluxUtil.java#L397) and context?

Reply:
Good catch!!, I'll update to use monoError.